### PR TITLE
SSL test framework: port resumption tests

### PR DESCRIPTION
**Pull request by ekasper**
Systematically test every server-side version downgrade or upgrade.

Client version upgrade or downgrade could be tested analogously but will
be done in a later change.

### CODE REVIEW

**Comment by ekasper** [comment]
Ping.

File: test/ssl_test_ctx.c [hunk 1]
```diff
@@ -280,6 +280,54 @@ const char *ssl_test_method_name(ssl_test_method_t method)
...
+
+const char *ssl_handshake_mode_name(ssl_handshake_mode_t mode)
+{
+    return enum_name(ssl_handshake_modes, OSSL_NELEM(ssl_handshake_modes),
+                     mode);
+}
+
+static int parse_boolean(const char *value, int *result)
+{
+    if (strcmp(value, "Yes") == 0) {
```
Comment (richsalz) [inline test/ssl_test_ctx.c]: strcasecmp?

**Comment by richsalz** [comment]
+1, with one nit.

**Comment by mattcaswell** [comment]
@ekasper Can this be merged now?

**Comment by ekasper** [comment]
@mattcaswell I just finished rebasing this. The tests started failing because I wasn't calling SSL_shutdown, and your commit e4612d02c53cccd24fa97b08fc01250d1238cca1 fixed a bug that let this go through previously.

I've added shutdown now so let's run this through Travis again - and also, can someone (you?) review the added shutdown code in handshake_helper.c?

File: test/ssl_test_ctx.c (hunk 1 above)
Comment (ekasper) [inline test/ssl_test_ctx.c]: done

File: test/generate_ssl_tests.pl [hunk 2]
```diff
@@ -43,11 +43,24 @@ sub print_templates {
...
-	if (defined $test->{"server2"}) {
+        if (defined $test->{"server2"}) {
             $test->{"server2"} = { (%ssltests::base_server, %{$test->{"server2"}}) };
+        } elsif (defined $test->{"test"}->{"ServerNameCallback"}) {
+            # Default is the same as server.
+            $test->{"server2"} = { (%ssltests::base_server, %{$test->{"server"}}) };
         } else {
-            $test->{"server2"} = { };
+            # Do not emit an empty "server2" section.
+            $test->{"server2"} = { }
```
Comment (mattcaswell) [inline test/generate_ssl_tests.pl]: Missing `;`?

File: test/generate_ssl_tests.pl [hunk 3]
```diff
@@ -43,11 +43,24 @@ sub print_templates {
...
+        }
+        if (defined $test->{"resume_server"}) {
+            $test->{"resume_server"} = { (%ssltests::base_server, %{$test->{"resume_server"}}) };
+        } elsif (defined $test->{"test"}->{"HandshakeMode"} &&
+                 $test->{"test"}->{"HandshakeMode"} eq "Resume") {
+            # Default is the same as server.
+            $test->{"resume_server"} = { (%ssltests::base_server, %{$test->{"server"}}) };
+        } else {
+            # Do not emit an empty "resume-server" section.
+            $test->{"resume_server"} = { }
```
Comment (mattcaswell) [inline test/generate_ssl_tests.pl]: and here

File: test/handshake_helper.c [hunk 4]
```diff
@@ -615,10 +656,45 @@ HANDSHAKE_RESULT *do_handshake(SSL_CTX *server_ctx, SSL_CTX *server2_ctx,
...
+                               SSL_CTX *client_ctx, SSL_CTX *resume_server_ctx,
+                               const SSL_TEST_CTX *test_ctx)
+{
+    HANDSHAKE_RESULT *result;
+    SSL_SESSION *session = NULL;
+
+    result = do_handshake_internal(server_ctx, server2_ctx, client_ctx,
+                                   test_ctx, NULL, &session);
+    if (test_ctx->handshake_mode == SSL_TEST_HANDSHAKE_SIMPLE)
+        goto err;
```
Comment (mattcaswell) [inline test/handshake_helper.c]: It's not really an `err` here is it? Would a different label name be better?

File: test/generate_ssl_tests.pl (hunk 2 above)
Comment (ekasper) [inline test/generate_ssl_tests.pl]: Apparently the last ; of a block is optional (you learn while you live...) but I have fixed it anyway for consistency.

Also meanwhile fixed a memleak I'd introduced during the rebase.

File: test/handshake_helper.c (hunk 4 above)
Comment (ekasper) [inline test/handshake_helper.c]: I somehow thought we were always supposed to use `err` for consistency but I don't know where I get that from. Happy to change it to `done`.

**Comment by richsalz** [comment]
For what it's worth:
```
; git grep -h goto | awk '{print $2;}' | sort | uniq -c | sort -n | tail -4
    171 f_err;
    185 opthelp;
   1544 end;
   4517 err;
```

File: test/handshake_helper.c (hunk 4 above)
Comment (mattcaswell) [inline test/handshake_helper.c]: Yes please.

**Comment by mattcaswell** [comment]
+1 subject to the label name nit being resolved.

File: test/generate_ssl_tests.pl (hunk 2 above)
Comment (levitte) [inline test/generate_ssl_tests.pl]: Yes.  In perl, the semi is a statement *separator*, not a statement *terminator* (like it is in C).

**Comment by levitte** [comment]
`git grep` ...  I've learned something new.  Thank you!

**Pull request closed without merging**